app/waf/src/controllers/adc.controller.ts

```diff
@@ -594,6 +595,88 @@ export class AdcController extends BaseController {
 
     return userDataB64Encoded;
   }
+
+  private async deleteOn(adc: Adc, addon: AddonReqValues): Promise<void> {
+    let reclaimFuncs: {[key: string]: Function} = {
+      license: async () => {
+        let doMgr = await OnboardingManager.instanlize(this.wafapp);
+        let doBody = await doMgr.assembleDo(adc, {onboarding: false});
+        this.logger.debug(
+          'Json used for revoke license: ' + JSON.stringify(doBody),
+        );
+        await doMgr.onboarding(doBody).then(async () => {
+          // TODO: create a unified bigipMgr
+          let mgmt = adc.management!;
+          let bigipMgr = await BigIpManager.instanlize({
+            username: mgmt.username,
+            password: mgmt.password,
+            ipAddr: mgmt.ipAddress,
+            port: mgmt.tcpPort,
+          });
+
+          let noLicensed = async () => {
+            return await bigipMgr.getLicense().then(license => {
+              return license.registrationKey === 'none';
+            });
+          };
+          await checkAndWait(noLicensed, 240).catch(() => {
+            throw new Error('Timeout for waiting for reclaiming license.');
+          });
+        });
+      },
+
+      network: async () => {
+        let networkMgr = await this.wafapp.get(WafBindingKeys.KeyNetworkDriver);
+        for (let network of Object.keys(adc.networks)) {
+          if (!adc.networks[network].ready) continue;
+
+          try {
+            let portId = adc.networks[network].portId!;
+            await networkMgr.deletePort(addon.userToken, portId).then(() => {
+              this.logger.debug(`Deleted port ${portId}`);
+              delete adc.networks[network].portId;
+              delete adc.networks[network].fixedIp;
+              delete adc.networks[network].macAddr;
+              adc.networks[network].ready = false;
+            });
+          } catch (error) {
+            this.logger.error(`delete port for ${network}: ${error.message}`);
+          }
+        }
+      },
+
+      vm: async () => {
+        let computeMgr = await this.wafapp.get(
+          WafBindingKeys.KeyComputeManager,
+        );
+        if (adc.compute.vmId) {
+          await computeMgr
+            .deleteServer(addon.userToken, adc.compute.vmId!, addon.tenantId)
+            .then(() => {
+              this.logger.debug(`Deleted the vm ${adc.compute.vmId!}`);
+              delete adc.compute.vmId;
+              adc.management = undefined;
```

Comment:
Will '{}' be better than 'undefined'?

Reply:
The properties of adc.management(ipaddr, port, user, pswd..) is not optional, which means, 
adc.management should have 2 types of values: undefined or not undefined(with properties in it.).

Reply:
I did a test just now. Adc.management can be '{}', or just have one or two properties. No problem occurs in the app.

Reply:
When I remove one property 'username', it tells me: 
```[ts] Property 'username' is missing in type '{ password: string; rootPass: string; tcpPort: number; ipAddress: string; }' but required in type '{ ipAddress: string; tcpPort: number; username: string; password: string; rootPass: string; }'.```

How did you test it?

Anyhow, why do you think {} is better than undefined?

I think that undefined means not set, while distinguishing {} and {...}, you have still check its properties..

Reply:
I use master branch code to create one adc object, for example, posting a '{}' in management property, and I can get that object back from db, whose management property is still '{}'.

Reply:
Maybe that should be a type checking and will only occur during compiling. The js code will never care that.

Reply:
The code we work on is typescript, which has strict type check. 
After it's compiled to nodejs, it becomes loose on type check. 
That's a core difference between ts and js. 

Anyhow, that's a little out of the topic mentioned here. I'd like to left it undefined here.